homeassistant/components/isy994/models.py

```diff
@@ -0,0 +1,133 @@
+
+    def uid_base(self, node: Node | Group | Variable | Program | NetworkCommand) -> str:
+        """Return the unique id base string for a given node."""
+        if isinstance(node, NetworkCommand):
+            return f"{self.uuid}_{CONF_NETWORK}_{node.address}"
+        return f"{self.uuid}_{node.address}"
+
+    @property
+    def unique_ids(self) -> set[tuple[Platform, str]]:
+        """Return all the unique ids for a config entry id."""
+        current_unique_ids: set[tuple[Platform, str]] = {
+            (Platform.BUTTON, f"{self.uuid}_query")
+        }
+
+        # Structure and prefixes here must match what's added in __init__ and helpers
+        for platform in NODE_PLATFORMS:
+            for node in self.nodes[platform]:
+                current_unique_ids.add((platform, self.uid_base(node)))
+
+        for platform in NODE_AUX_PROP_PLATFORMS:
+            for node, control in self.aux_props[platform]:
+                current_unique_ids.add((platform, f"{self.uid_base(node)}_{control}"))
+
+        for platform in PROGRAM_PLATFORMS:
+            for _, node, _ in self.programs[platform]:
+                current_unique_ids.add((platform, self.uid_base(node)))
+
+        for platform in VARIABLE_PLATFORMS:
+            for node in self.variables[platform]:
+                current_unique_ids.add((platform, self.uid_base(node)))
+                if platform == Platform.NUMBER:
+                    current_unique_ids.add((platform, f"{self.uid_base(node)}_init"))
+
+        for platform in ROOT_NODE_PLATFORMS:
+            for node in self.root_nodes[platform]:
+                current_unique_ids.add((platform, f"{self.uid_base(node)}_query"))
+                if platform == Platform.BUTTON and node.protocol == PROTO_INSTEON:
+                    current_unique_ids.add((platform, f"{self.uid_base(node)}_beep"))
+
+        for node in self.net_resources:
+            current_unique_ids.add((Platform.BUTTON, self.uid_base(node)))
+
+        return current_unique_ids
+
+    @callback
+    def async_cleanup_registry_entries(self) -> None:
+        """Remove extra entities that are no longer part of the integration."""
+        entity_registry = er.async_get(self.hass)
+
+        existing_entries = er.async_entries_for_config_entry(
+            entity_registry, self.config_entry.entry_id
+        )
+        entities = {
+            (entity.domain, entity.unique_id): entity.entity_id
+            for entity in existing_entries
+        }
+
+        extra_entities = set(entities.keys()).difference(self.unique_ids)
+        if not extra_entities:
+            return
+
+        for entity in extra_entities:
+            if entity_registry.async_is_registered(entities[entity]):
+                entity_registry.async_remove(entities[entity])
+
+        _LOGGER.debug(
+            ("Cleaning up ISY entities: removed %s extra entities for %s"),
+            len(extra_entities),
+            self.config_entry.title,
+        )
```

Comment:
I think it would be cleaner to not put this in the dataclass and leave it as an external function which would also allow dropping of `hass` here

Reply:
Moved to util